kubeflow/application/README.md

```diff
@@ -0,0 +1,137 @@
+  creationTimestamp: 2018-12-20T21:40:53Z
+  generation: 1
+  labels:
+    app: kubeflow
+    app.kubernetes.io/name: kubeflow
+    ksonnet.io/component: application
+  name: kubeflow
+  namespace: kubeflow
+  resourceVersion: "2889700"
+  selfLink: /apis/app.k8s.io/v1beta1/namespaces/kubeflow/applications/kubeflow
+  uid: ed56d218-049f-11e9-88a0-42010a8a01a3
+spec:
+  assemblyPhase: Succeeded
+  componentKinds:
+  - group: rbac.authorization.k8s.io/v1beta1
+    kind: ClusterRoleBinding
+  - group: rbac.authorization.k8s.io/v1beta1
+    kind: ClusterRole
+  - group: v1
+    kind: ConfigMap
+  - group: apiextensions.k8s.io/v1beta1
+    kind: CustomResourceDefinition
+  - group: apps/v1beta1
+    kind: Deployment
+  - group: apps/v1beta2
+    kind: Deployment
+  - group: extensions/v1beta1
+    kind: Deployment
+  - group: batch/v1
+    kind: Job
+  - group: v1
+    kind: PersistentVolumeClaim
+  - group: rbac.authorization.k8s.io/v1beta1
+    kind: RoleBinding
+  - group: rbac.authorization.k8s.io/v1beta1
+    kind: Role
+  - group: v1
+    kind: Secret
+  - group: v1
+    kind: ServiceAccount
+  - group: v1
+    kind: Service
+  - group: apps/v1beta1
+    kind: StatefulSet
+  descriptor:
+    description: ""
+    icons: []
+    keywords: []
+    links: []
+    maintainers: []
+    notes: ""
+    owners: []
+    type: kubeflow
+    version: "0.4"
+  info: []
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: kubeflow
+```
+
+Alternatively, the application component can produce an Application Kind based on a subset 
+of the generated components by setting the components parameter. For example
+
+```bash
+ks param set application components '["ambassador","jupyter"]'
+```
+
+## Options (emitCRD=true, emitController=false)
+
+- emitCRD (=true)
+If the Application CRD has already been emitted then set this to false
+
+```
+ks param set application emitCRD false
+```
+
+- emitController (=false)
+
+If this flag is true then the components normally deployed directly to the api-server will be emitted 
+by the application-controller in-cluster. 
+
+## Example Script (Deploying components '["ambassador","jupyter","centraldashboard","tensorboard","tf-training"]')
+
+```
+#!/usr/bin/env bash
+KUBEFLOW_DIR=$HOME/go/src/github.com/kubeflow/kubeflow
+
+cd $HOME
+if [[ -d kf_app ]]; then
+  rm -rf kf_app
+fi
+kubectl get ns -oname | grep kubeflow 2>&1>/dev/null
+if (( $? == 0 )); then
+  kubectl delete ns kubeflow
+fi
+
+$KUBEFLOW_DIR/scripts/kfctl.sh init kf_app --platform none
+cd $HOME/kf_app
+$KUBEFLOW_DIR/scripts/kfctl.sh generate all
+cd $HOME/kf_app/ks_app
+kubectl create ns kubeflow
+ks env add default --namespace kubeflow
+ks param set application name kubeflow
+ks param set application emitController true
+ks param set application components '["ambassador","jupyter","centraldashboard","tensorboard","tf-training"]'
```

Comment:
Why isn't generating the Application resource(s) describing Kubeflow components done automatically by kfctl?

Is this because of the problem discussed in a different PR that kfctl.sh doesn't know which applications are installed?

What if we don't treat Kubeflow as one monolithic application but as a discrete set of applications.
e.g. ambassador, jupyter, centraldashboard, tensorboard, tf-training are each their own applications each with its own Application resource describing it. Each application resource would then be one of the resources created by the corresponding KS component.

Reply:
Yes we could do it this way - using kfctl (golang). Though dependencies are no longer part of the Application CRD but are in the [Road Map](https://github.com/kubernetes-sigs/application/issues/33)